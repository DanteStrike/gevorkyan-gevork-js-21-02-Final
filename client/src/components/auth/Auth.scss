@import "src/assets/variables";

.auth {
  display: flex;
  align-items: center;
  justify-content: flex-end;
  width: 180px;
  height: 50px;

  &__link {
    padding: 0 0 0 8px;
    font-family: var(--secondary-font-family);
    font-size: 18px;
    color: inherit;
  }

  &__link--left {
    padding-right: 8px;
    border-right: var(--primary-border);
  }

  &__link--small {
    font-size: 12px;
    color: var(--secondory-font-color);
  }

  &__btn {
    padding: 4px 0 0 8px;
    font-family: var(--secondary-font-family);
    font-size: 12px;
    color: var(--secondory-font-color);
    cursor: pointer;
    background: unset;
    border: none;
    transition: color 0.3s;
  }

  &__btn:hover {
    color: #1890ff;
  }

  &__link--profile {
    display: flex;
    flex-grow: 1;
    align-items: center;
    justify-content: flex-end;
  }

  &__user {
    max-width: 108px;
    margin-left: 5px;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }
}

@media (max-width: $tabletWidth) {
  .auth {
    justify-content: center;
    height: 40px;

    &__link {
      padding: 0 8px;
      font-size: 16px;
    }
  }
}
