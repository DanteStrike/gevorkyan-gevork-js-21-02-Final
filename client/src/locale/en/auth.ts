const AUTH = {
  login: `Login`,
  registration: `Registration`,
  logout: `Logout`,
};

export default AUTH;
