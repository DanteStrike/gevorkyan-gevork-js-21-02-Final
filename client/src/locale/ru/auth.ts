const AUTH = {
  login: `Вход`,
  registration: `Регистрация`,
  logout: `Выход`,
};

export default AUTH;
