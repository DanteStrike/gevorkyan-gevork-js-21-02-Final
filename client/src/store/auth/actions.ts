import types, {authStorageKey} from './types';
import {fetchActions} from './slices/fetch';

const login = (id: string, name: string, picture: string) => ({
  type: types.LOGIN,
  payload: {
    id,
    picture,
    name,
  },
});

const logout = () => {
  localStorage.setItem(authStorageKey, ``);

  return {
    type: types.LOGOUT,
  };
};

const actions = {
  login,
  logout,
  ...fetchActions,
};

export default actions;
