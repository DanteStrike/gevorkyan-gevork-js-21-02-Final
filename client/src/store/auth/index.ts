export {default as authActions} from './actions';
export {default as authOperations} from './operations';
export {default as authSelectors} from './selectors';
export {default as authReducer} from './reducers';
