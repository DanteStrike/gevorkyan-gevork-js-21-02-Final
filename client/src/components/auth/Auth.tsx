import React from 'react';
import './Auth.scss';
import {Avatar} from 'antd';
import CustomLink from '../custom-link/CustomLink';
import {RoutePath} from '../../enums';
import {RouteUtils} from '../../utils';
import useAppTranslation from '../../hooks/use-app-translation';

interface IAuthProps {
  isAuth: boolean;
  onLogout: () => void;
  authUser?: {
    id: string;
    name: string;
    picture: string;
  };
}

function Auth({
  isAuth,
  onLogout,
  authUser = {
    id: ``,
    name: ``,
    picture: ``,
  },
}: IAuthProps) {
  const {t} = useAppTranslation(`auth`);
  const {name, picture, id} = authUser;
  const profileRoute = RouteUtils.createProfileRoute(id);

  return (
    <div className="auth">
      {!isAuth ? (
        <>
          <CustomLink className="auth__link auth__link--left" to={RoutePath.LOGIN}>
            {t(`login`)}
          </CustomLink>
          <CustomLink className="auth__link" to={RoutePath.REGISTRATION}>
            {t(`registration`)}
          </CustomLink>
        </>
      ) : (
        <>
          <CustomLink className="auth__link auth__link--left auth__link--profile" to={profileRoute}>
            <Avatar src={picture} />
            <span className="auth__user">{name}</span>
          </CustomLink>
          <button type="button" className="auth__btn" onClick={onLogout}>
            {t(`logout`)}
          </button>
        </>
      )}
    </div>
  );
}

Auth.defaultProps = {
  authUser: {
    id: ``,
    name: ``,
    picture: ``,
  },
};

export default React.memo(Auth);
