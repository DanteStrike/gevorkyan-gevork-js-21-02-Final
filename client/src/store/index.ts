import {applyMiddleware, combineReducers, createStore} from 'redux';
import thunk from 'redux-thunk';
import {composeWithDevTools} from 'redux-devtools-extension';
import configureAPI from '../server/cofigure-api';
import {usersReducer} from './users';
import NameSpace from './name-space';
import {profileReducer} from './profile';
import {postsReducer} from './posts';
import {authReducer} from './auth';
import {postReducer} from './post';

const api = configureAPI();
const rootReducer = combineReducers({
  [NameSpace.USERS]: usersReducer,
  [NameSpace.PROFILE]: profileReducer,
  [NameSpace.POSTS]: postsReducer,
  [NameSpace.AUTH]: authReducer,
  [NameSpace.POST]: postReducer,
});

const configuredStore = createStore(rootReducer, composeWithDevTools(applyMiddleware(thunk.withExtraArgument(api))));
export type RootState = ReturnType<typeof configuredStore.getState>;
export type AppDispatch = typeof configuredStore.dispatch;
export default configuredStore;
