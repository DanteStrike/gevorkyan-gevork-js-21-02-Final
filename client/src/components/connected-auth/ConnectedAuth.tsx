import React from 'react';
import Auth from '../auth/Auth';
import useAppSelector from '../../hooks/use-app-selector';
import {authActions, authSelectors} from '../../store/auth';
import useAppDispatch from '../../hooks/use-app-dispatch';

function ConnectedAuth() {
  const dispatch = useAppDispatch();
  const isAuth = useAppSelector(authSelectors.getIsAuth);
  const authUser = useAppSelector(authSelectors.getAuthUser);

  const handleLogout = () => {
    dispatch(authActions.logout());
  };

  return (
    <>
      <Auth isAuth={isAuth} authUser={authUser} onLogout={handleLogout} />
    </>
  );
}

export default React.memo(ConnectedAuth);
