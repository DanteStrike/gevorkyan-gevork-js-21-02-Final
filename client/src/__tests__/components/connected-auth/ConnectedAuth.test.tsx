import {mount, shallow} from 'enzyme';
import React from 'react';
import {Provider} from 'react-redux';
import {MemoryRouter} from 'react-router-dom';
import ConnectedAuth from '../../../components/connected-auth/ConnectedAuth';
import mockInitStore from '../../mocks/store.mock';
import {authActions} from '../../../store/auth';

describe(`Component ConnectedAuth should render correctly`, () => {
  it(`snapshot default test`, () => {
    const wrapper = shallow(
      <Provider store={mockInitStore}>
        <ConnectedAuth />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe(`Component ConnectedAuth should work correctly`, () => {
  it(`logout test`, () => {
    const wrapper = mount(
      <MemoryRouter>
        <Provider store={mockInitStore}>
          <ConnectedAuth />
        </Provider>
      </MemoryRouter>
    );
    (wrapper.find(`Auth`).prop('onLogout') as Function)();
    expect(mockInitStore.dispatch).toHaveBeenCalledTimes(1);
    expect(mockInitStore.dispatch).toHaveBeenLastCalledWith(authActions.logout());
  });
});
