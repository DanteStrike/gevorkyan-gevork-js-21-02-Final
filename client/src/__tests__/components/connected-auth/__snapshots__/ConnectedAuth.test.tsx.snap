// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component ConnectedAuth should render correctly snapshot default test 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "onStateChange": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <Memo(ConnectedAuth) />
</ContextProvider>
`;
